File: backend/src/inventoryview/schemas/usage.py
"""Usage analytics and login audit request/response schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


# -- Event ingestion --

class UsageEventRequest(BaseModel):
    """Single usage event submission."""

    feature_area: str = Field(max_length=100)
    action: str = Field(max_length=100)


class UsageEventBatchRequest(BaseModel):
    """Batch usage event submission."""

    events: list[UsageEventRequest] = Field(max_items=50)


class UsageEventResponse(BaseModel):
    """Response after recording usage event(s)."""

    status: str = "ok"
    count: int | None = None


# -- Dashboard summary --

class Period(BaseModel):
    """Time period for usage queries."""

    start: datetime
    end: datetime


class FeatureAreaSummary(BaseModel):
    """Aggregate stats for a single feature area."""

    feature_area: str
    total_events: int
    unique_users: int
    trend: str  # "up", "down", "flat"
    trend_percentage: float


class UsageSummaryResponse(BaseModel):
    """Top-level usage dashboard response."""

    period: Period
    feature_areas: list[FeatureAreaSummary]
    total_events: int
    total_unique_users: int


# -- Feature detail --

class ActionBreakdown(BaseModel):
    """Stats for a single action within a feature area."""

    action: str
    count: int
    unique_users: int


class FeatureDetailResponse(BaseModel):
    """Detailed action breakdown for a feature area."""

    feature_area: str
    period: Period
    actions: list[ActionBreakdown]
    total_events: int


# -- Login audit --

class LoginAuditEntry(BaseModel):
    """Single login audit log entry."""

    id: UUID
    username: str
    outcome: str
    failure_reason: str | None = None
    ip_address: str
    created_at: datetime


class LoginSummary(BaseModel):
    """Summary counts for login activity."""

    total_attempts: int
    successful: int
    failed: int
    unique_users: int


class LoginAuditResponse(BaseModel):
    """Paginated login audit response."""

    period: Period
    summary: LoginSummary
    entries: list[LoginAuditEntry]
    page: int
    page_size: int
    total_count: int

File: frontend/src/api/usage.ts
import apiClient from "./client";

// -- Types --

export interface Period {
  start: string;
  end: string;
}

export interface FeatureAreaSummary {
  feature_area: string;
  total_events: number;
  unique_users: number;
  trend: "up" | "down" | "flat";
  trend_percentage: number;
}

export interface UsageSummaryResponse {
  period: Period;
  feature_areas: FeatureAreaSummary[];
  total_events: number;
  total_unique_users: number;
}

export interface ActionBreakdown {
  action: string;
  count: number;
  unique_users: number;
}

export interface FeatureDetailResponse {
  feature_area: string;
  period: Period;
  actions: ActionBreakdown[];
  total_events: number;
}

export interface LoginAuditEntry {
  id: string;
  username: string;
  outcome: "success" | "failure";
  failure_reason: string | null;
  ip_address: string;
  created_at: string;
}

export interface LoginSummary {
  total_attempts: number;
  successful: number;
  failed: number;
  unique_users: number;
}

export interface LoginAuditResponse {
  period: Period;
  summary: LoginSummary;
  entries: LoginAuditEntry[];
  page: number;
  page_size: number;
  total_count: number;
}

// -- API calls --

export async function trackEvent(featureArea: string, action: string): Promise<void> {
  await apiClient.post("/usage/events", { feature_area: featureArea, action });
}

export async function trackEventsBatch(
  events: { feature_area: string; action: string }[],
): Promise<void> {
  await apiClient.post("/usage/events/batch", { events });
}

export async function getUsageSummary(
  startDate?: string,
  endDate?: string,
): Promise<UsageSummaryResponse> {
  const params: Record<string, string> = {};
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const res = await apiClient.get<UsageSummaryResponse>("/usage/summary", { params });
  return res.data;
}

export async function getFeatureDetail(
  featureArea: string,
  startDate?: string,
  endDate?: string,
): Promise<FeatureDetailResponse> {
  const params: Record<string, string> = {};
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const res = await apiClient.get<FeatureDetailResponse>(
    `/usage/feature/${encodeURIComponent(featureArea)}`,
    { params },
  );
  return res.data;
}

export async function getLoginAudit(
  startDate?: string,
  endDate?: string,
  page = 1,
  pageSize = 50,
): Promise<LoginAuditResponse> {
  const params: Record<string, string | number> = { page, page_size: pageSize };
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const res = await apiClient.get<LoginAuditResponse>("/usage/logins", { params });
  return res.data;
}
